File: src/main/java/com/learning/leetcode/arrays/medium/MaxPointsFromCards.java
package com.learning.leetcode.arrays.medium;

import java.util.Arrays;

public class MaxPointsFromCards {
    public static void main(String[] args) {
        int[] nums = {1,2,3,4,5,6,1};
        int k = 3;
        System.out.println("Given Array: " + Arrays.toString(nums));
        System.out.println("Max score obtained in " + k + " steps is " + maxScore(nums, k));
        nums = new int[]{2,2,2};
        k = 2;
        System.out.println("Given Array: " + Arrays.toString(nums));
        System.out.println("Max score obtained in " + k + " steps is " + maxScore(nums, k));
        nums = new int[]{9,7,7,9,7,7,9};
        k = 7;
        System.out.println("Given Array: " + Arrays.toString(nums));
        System.out.println("Max score obtained in " + k + " steps is " + maxScore(nums, k));
    }

    private static int maxScore(int[] cardPoints, int k) {
        int left = 0, right = cardPoints.length - k;
        int sum=0;
        for(int i = right; i < cardPoints.length; i++){
            sum += cardPoints[i];
        }
        int result = sum;
        while(right < cardPoints.length){
            sum += (cardPoints[left] - cardPoints[right]);
            result = Math.max(result, sum);
            right++;
            left++;
        }
        return result;
    }
}

File: src/main/java/com/learning/leetcode/arrays/medium/MaxPointsFromCards.py
from typing import List


def maxScore(cardPoints: List[int], k: int) -> int:
    l, r = 0, len(cardPoints) - k
    total = sum(cardPoints[r:])
    result = total
    while r < len(cardPoints):
        total += (cardPoints[l] - cardPoints[r])
        result = max(result, total)
        l += 1
        r += 1
    return result


def main():
    nums = [1, 2, 3, 4, 5, 6, 1]
    k = 3
    print("Given Array: ", nums)
    print("Max points obtained for %d cards is: %d " % (k, maxScore(nums, k)))
    nums = [2, 2, 2]
    k = 2
    print("Given Array: ", nums)
    print("Max points obtained for %d cards is: %d " % (k, maxScore(nums, k)))
    nums = [9, 7, 7, 9, 7, 7, 9]
    k = 7
    print("Given Array: ", nums)
    print("Max points obtained for %d cards is: %d " % (k, maxScore(nums, k)))


main()
